### PR TITLE
Add defensive check for member accesses

### LINKED ISSUE #1757: Add defensive check for member access
### Issue To Be Solved

Bugs in the type checker and interpreter may lead to security issues. 

### Suggested Solution

Add an additional defensive check which validates that the type which is performed a member-access on matches the static type of the member expression.

### PR DESCRIPTION
**Pull request by turbolent**
Closes #1757 

## Description

Defensively check if the value's type of a member access matches the expected static type.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/errors.go [hunk 1]
```diff
@@ -510,19 +510,17 @@ func (e InvocationArgumentTypeError) Error() string {
 	)
 }
 
-// InvocationReceiverTypeError
+// MemberAccessTypeError
 //
-type InvocationReceiverTypeError struct {
```
Comment (turbolent) [inline runtime/interpreter/errors.go:515]: This error was a leftover from a temporary period where we defensively checked an invocation's receiver (`v` in `v.f()`).

This PR is a generalization to any member access, reads and writes (`v.p` and `v.p = ...`), which includes also reads which are then invoked (`v.f()`).

File: runtime/tests/interpreter/reference_test.go [hunk 2]
```diff
@@ -472,9 +472,11 @@ func TestInterpretResourceReferenceAfterMove(t *testing.T) {
 		)
```
Comment (turbolent) [inline runtime/tests/interpreter/reference_test.go:472]: The new check uncovered some incorrect tests

File: runtime/tests/interpreter/memory_metering_test.go [hunk 3]
```diff
@@ -8713,7 +8713,7 @@ func TestInterpretFunctionStaticType(t *testing.T) {
 		_, err := inter.Invoke("main")
 		require.NoError(t, err)
 
-		assert.Equal(t, uint64(1), meter.getMemory(common.MemoryKindFunctionStaticType))
+		assert.Equal(t, uint64(2), meter.getMemory(common.MemoryKindFunctionStaticType))
```
Comment (dsainati1) [inline runtime/tests/interpreter/memory_metering_test.go:8716]: Why has this increased? Does this defensive check allocate another static type?

Comment (turbolent) [reply runtime/tests/interpreter/memory_metering_test.go:8716]: Yes, the defensive check gets the static type of the value which the member access is performed on

**Review by dsainati1 — APPROVED** [review]

**Review by robert-e-davidson3 — APPROVED** [review]

File: runtime/interpreter/interpreter_expression.go [hunk 4]
```diff
@@ -177,11 +182,51 @@ func (interpreter *Interpreter) memberExpressionGetterSetter(memberExpression *a
...
+	case *sema.TransactionType:
+		// TODO: maybe also check transactions.
+		//   they are composites with a type ID which has an empty qualified ID, i.e. no type is available
+
+		return
+
+	case *sema.CompositeType:
+		// TODO: also check built-in values.
+		//   blocked by standard library values (RLP, BLS, etc.),
+		//   which are implemented as contracts, but currently do not have their type registered
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/interpreter/interpreter_expression.go:210]: Yes, we should register the types from the standard library. Elaborations are for user programs, one per location (e.g contract). Given that these types are built-in, we would need to register the types in a special way.

File: runtime/tests/interpreter/member_test.go [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+                    fun get(s: S) {
+                        s.foo
+                    }
+
+                    fun set(s: S) {
+                        s.foo = 2
+                    }
+                `)
+
+				value, err := inter.Invoke("S")
```
Comment (SupunS) [inline runtime/tests/interpreter/member_test.go:60]: What does invoking "S" do? Does it directly calls the Constructor of "S"?

**Review by SupunS — APPROVED** [review]

File: runtime/tests/interpreter/member_test.go (hunk 5 above)
Comment (turbolent) [reply runtime/tests/interpreter/member_test.go:60]: Yes, declaring a composite declares a value for the constructor, a function with the name of the type. I'll add a comment

**Pull request merged**